cumulusci/utils/clean_invalid_references.py

```diff
@@ -0,0 +1,365 @@
+    "fields": [FIELD_PERMISSION],
+    "tabs": [TAB_PERMISSION],
+    "applications": [APP_PERMISSION],
+    "customPermissions": [CUSTOMPERM_PERMISSION],
+    "pages": [APEXPAGE_PERMISSION],
+    "classes": [APEXCLASS_PERMISSION],
+    "flows": [FLOW_PERMISSION],
+    "recordTypes": [RECORDTYPE_PERMISSION],
+    "userPermissions": [USER_PERMISSION],
+}
+
+
+def run_queries_in_parallel(
+    queries: Dict[str, str], run_query: Callable[[str], dict], num_threads: int = 4
+):
+    """Accepts a set of queries structured as {'query_name': 'query'}
+    and a run_query function that runs a particular query. Runs queries in parallel and returns the queries"""
+    results_dict = {}
+
+    with ThreadPoolExecutor(max_workers=num_threads) as executor:
+        futures = {
+            query_name: executor.submit(run_query, query)
+            for query_name, query in queries.items()
+        }
+
+    for query_name, future in futures.items():
+        try:
+            query_result = future.result()
+            results_dict[query_name] = query_result
+        except Exception as e:
+            raise Exception(f"Error executing query '{query_name}': {type(e)}: {e}")
+        else:
+            queries.pop(query_name, None)
+
+    return results_dict
+
+
+def entities_from_query(sf: Salesforce):
+    # Define the query
+    def run_query(path: str):
+        urlpath = sf.base_url + path
+        method = "GET"
+        return sf._call_salesforce(method, urlpath).json()
+
+    # Queries
+    queries = {}
+    queries["userPermissions"] = "sobjects/PermissionSet/describe"
+    queries["fields"] = "sobjects/FieldPermissions/describe"
+    queries["objects"] = "sobjects/ObjectPermissions/describe"
+    queries["tabs"] = "query/?q=SELECT+Name+FROM+PermissionSetTabSetting+GROUP+BY+Name"
+
+    # Run all queries
+    result = run_queries_in_parallel(queries, run_query)
+
+    # Process the results
+    user_permissions = process_user_permissions(result["userPermissions"])
+    fields = process_fields(result["fields"])
+    tabs = process_tabs(result["tabs"])
+    objects = process_objects(result["objects"])
+
+    return user_permissions, fields, tabs, objects
+
+
+def process_user_permissions(result_dict: dict) -> set:
+    permissions = [
+        f["name"][len("Permissions") :]
+        for f in result_dict["fields"]
+        if f["name"].startswith("Permissions") and f["type"] == "boolean"
+    ]
+    return set(permissions)
+
+
+def process_fields(result_dict: dict) -> set:
+    field_entities = []
+    for field in result_dict["fields"]:
+        if field.get("name") == "Field":
+            field_entities.extend(
+                picklistValue["value"]
+                for picklistValue in field.get("picklistValues", [])
+            )
+    return set(field_entities)
+
+
+def process_tabs(result_dict: dict) -> set:
+    tabs = [tab["Name"] for tab in result_dict["records"]]
+    return set(tabs)
```

Comment:
Fun Python trick. You can use a _set comprehension_ here:

```suggestion
    return {tab["Name"] for tab in result_dict["records"]}
```

Curly braces are a set comprehension unless you have a `k:v` structure, which makes it a dict comprehension.

Reply:
Yep, will do that for the rest of the functions too!